test/server/name_bazaar/smart_contract_tests.cljs

```diff
@@ -496,12 +496,14 @@
                          (auction-offering/finalize! offering {:from addr0}))))
           (testing "Ensuring the new owner gets his name"
             (is (= addr3 (ens/owner {:ens.record/node (namehash "abc.eth")}))))
-          (testing "Ensuring the new owner gets his deed"
-            (is (= addr3 (registrar/entry-deed-owner {:ens.record/label "abc"}))))
+          (testing "Ensuring the new owner gets his registration"
+            (is (= addr3 (registrar/registration-owner {:ens.record/label "abc"}))))
 
           (testing "Ensuring the previous owner gets the funds"
-            (let [bid-tx-1 (:t1-register @transaction-log)
-                  tx-total-cost (get-transaction-cost bid-tx-1)
+            (let [bid-1-tx-1 (:t1-transfer-tx1 @transaction-log)
+                  bid-1-tx-2 (:t1-transfer-tx2 @transaction-log)
```

Comment:
Why there are two txs here?

Reply:
Ownership transfer now does two transactions instead of one. As mentioned above, otherwise we'd need to be supporting reclaim call to fully get your name, which would be a second transaction anyway.

Reply:
Oh, that's because `doTransferOwnership` in `Offering.sol` changed and we use 2 contract calls right?

Reply:
Not really. Referenced diff happened because `registrar/transfer!` changed. When we (the guy offering his domain) transfer our name ownership to the offering contract, we're creating two transactions. On the other hand, when the offering contract passes the stored name to the next owner on auction finalization, it happens in one transaction (e.g. on finalize call). So the two actions of name transfer (registry transfer and registrar/token transfer) don't always happen in same amount of transactions.

Reply:
I see. That's basically the `transferOwnership` function from `Offering.sol` being called after finalize.

But basically you just want to call `doTransferOwnership` from ens no? You don't always need to call `transferFrom`, that is only for TLDs no?

Reply:
Aha, you very correctly spotted that `registrar/transfer!` is missing a condition. Thanks!

Reply:
Maybe create a new function in solidity (that internally calls the private `doTransferOwnership` fn)? I think it will save some gas for users to call just a single tx.

Reply:
`doTransferOwnership` transfers ownership away from the Offering contract, and that's possible because the Offering contract already owns the name. You can't make it give someone's ownership to itself. In `registrar/transfer!`'s calls it's the owning account that's instructing the ownership changes. If you want to allow someone else change ownerships in your stead in Registry resp. Registrar, then you need to set them as operator resp. controller. By the way, complicating the offering contract would cost you too as it's deployed for each offering.

Reply:
Ok so it seems it was not wise of me to assume I can completely ignore UI layer in this AT ALL.

When you work through `(defn transfer-ownership-button` you should find that `registrar/transfer!` is not used in it, and probably not anywhere else in codebase outside of tests (I was always surprised by lack of `registrar/transfer!` calls, but I thought it's maybe being called in some opaque way or it's all channeling through the use in `generator.cljs`, which I now find nonsense). Further, the logic of figuring out whether we're tld or not and then doing a corresponding transfer resides completely in the UI layer files, which I find frankly astounding.

Reply:
Actually, the `generator.cljs` seems intersting for development puroposes. I would say the point of it is to privide some sample data to work with while in dev mode.

> Further, the logic of figuring out whether we're tld or not and then doing a corresponding transfer resides completely in the UI layer files, which I find frankly astounding.

What do you mean by this? Whether a domain is TLD is trivial and can be done on FE, but I guess you mean something else.

Reply:
I mean e.g. creating transactions. Explore `:onClick`s of buttons and other UI entrypoints to see how things get done.

Look at
```
(reg-event-fx
  :name-bazaar-registrar/transfer
```
That's around where I'd expect a contract_api call (`registrar/transfer!`), but you get the thing done independently via `:district0x/make-transaction`.

Reply:
Thanks, I think I understand.